### PR TITLE
[scan] flatten subgraph output and make subgraph inputs to be a slice

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135603
* #137967
* #135602
* __->__ #135601
* #135600

This pr introduces two changes:
1. Before this pr, the subgraphs output is ([], []), in this pr, we change it to a flattened list for easier codegen and consistency with other control flow operators.

2. Before the PR, the combine_fn of scan takes a sliced input but keep the sliced dimension. For exmaple, suppose xs = torch.randn(3, 4, 5) and we scan over dim 0, the combine_fn looks like:
```
# x.shape = (1, 4, 5) instead of (4, 5)
def combine_fn(carry, x):
  ...
```

In this PR, we fixed this and also simplify some of the slicing logic.

3. this diff also make sure we always stack ys on fist dimension.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: test/functorch/test_control_flow.py [hunk 1]
```diff
@@ -33,6 +33,15 @@
 )
 
 
+# say we have a tensor of shape [3, 4, 5, 6]
+# shift_first_dim_to(t, 3) -> [4, 5, 6, 3]
+def shift_first_dim_to(t, to_dim: int):
+    assert to_dim >= 0 and to_dim < t.ndim
+    sz_list = list(t.size())
+    dims = list(range(1, to_dim + 1)) + [0] + list(range(to_dim + 1, t.ndim))
+    return t.permute(*dims)
```
Comment (zou3519) [inline test/functorch/test_control_flow.py:42]: use t.movedim(1, 3)

File: torch/_higher_order_ops/scan.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # mypy: allow-untyped-defs
 import functools
 import itertools
-from typing import Callable, List, Tuple
+from typing import Any, Callable, List, Tuple
```
Comment (zou3519) [inline torch/_higher_order_ops/scan.py:4]: For future reference: it sounds like the PR could have been split into "make subgraph inputs be a a slice" and "flatten subgraph output", it's a bit difficult for me to disentangle this

**Review by zou3519 — APPROVED** [review]

File: test/functorch/test_control_flow.py (hunk 1 above)
Comment (ydwu4) [reply test/functorch/test_control_flow.py:44]: wow! Thanks for pointing it out!

File: torch/_higher_order_ops/scan.py (hunk 2 above)
Comment (ydwu4) [reply torch/_higher_order_ops/scan.py:4]: okok

**Pull request closed without merging**